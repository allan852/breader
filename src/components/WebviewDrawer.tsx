import { Drawer, Icon } from 'antd'
import { shell, WebviewTag } from 'electron'
import React, { createRef, PureComponent, RefObject } from 'react'
import ProgressBar from './ProgressBar'

import '../styles/WebviewDrawer.less'

export interface IWebviewDrawerProps {
  visible: boolean
  src: string
  width: string | number
  onClose: (e: any) => any
}
interface IWebviewDrawerState {
  hasProgressBar: boolean
}

class WebviewDrawer extends PureComponent<
  IWebviewDrawerProps,
  IWebviewDrawerState
> {
  public webview?: WebviewTag
  public openTimes: number
  public userAgent =
    'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1'
  private _progressBar: RefObject<ProgressBar>
  private _needUpdateWebview: boolean
  public constructor(props: IWebviewDrawerProps) {
    super(props)
    this.state = { hasProgressBar: false }
    this._progressBar = createRef<ProgressBar>()
    this._needUpdateWebview = false
    this.openTimes = 0
  }
  public handleCompassClick = () => {
    if (this.props.src) {
      shell.openExternal(this.props.src)
    }
  }
  public handleDrawerClose = (e: any) => {
    this.props.onClose(e)
    const times = this.openTimes
    const webview = this.webview
    if (webview) {
      const src = webview.src
      setTimeout(() => {
        if (
          this.webview &&
          this.openTimes === times &&
          this.props.visible === false &&
          webview === this.webview &&
          this.webview.src === src
        ) {
          this.webview.remove()
          this.webview = undefined
        }
      }, 10 * 1000)
    }
  }
  public makeWebView(url: string) {
    if (this.webview) {
      this.webview.src = url
    } else {
      const div = document.querySelector('.drawer-content')
      if (div) {
        this.webview = document.createElement('webview')
        this.webview.addEventListener('did-start-loading', this.showProgressBar)
        this.webview.addEventListener('did-stop-loading', this.hideProgressBar)
        this.webview.src = url
        this.webview.useragent = this.userAgent
        div.appendChild(this.webview)
      }
    }
  }
  public showProgressBar = () => {
    this.setState({ hasProgressBar: true })
  }
  public hideProgressBar = () => {
    const bar = this._progressBar.current
    if (bar) {
      bar.toEnd()
    }
  }
  public handleProgressBarEnd = () => {
    this.setState({ hasProgressBar: false })
  }
  public getSnapshotBeforeUpdate(prevProps: IWebviewDrawerProps) {
    // for the legacy componentWillReceiveProps
    const props = this.props
    if (props.visible) {
      if (!prevProps.visible || prevProps.src !== props.src) {
        this.openTimes++
        if (!this.webview || this.webview.src !== props.src) {
          this._needUpdateWebview = true
        }
      }
    }
    return null
  }
  public componentDidUpdate(prevProps: IWebviewDrawerProps) {
    const props = this.props
    if (this._needUpdateWebview) {
      this._needUpdateWebview = false
      this.setState({
        hasProgressBar: true,
      })
      setImmediate(() => {
        this.makeWebView(props.src)
      })
    }
  }
  public componentWillUnmount() {
    this.setState = () => null
  }
  public render() {
    return (
      <Drawer
        className="webview-drawer"
        closable={false}
        visible={this.props.visible}
        onClose={this.handleDrawerClose}
        width={this.props.width}>
        <div className="drawer-header">
          <div className="drawer-header-left" onClick={this.handleDrawerClose}>
            <Icon type="close" />
          </div>
          <div
            className="drawer-header-right"
            onClick={this.handleCompassClick}>
            <Icon type="compass" />
          </div>
          {this.state.hasProgressBar && (
            <ProgressBar
              ref={this._progressBar}
              className="webview-progress"
              time={6}
              max={90}
              onEnd={this.handleProgressBarEnd}
            />
          )}
        </div>
        <div className="drawer-content">
          {/* <webview ref={this.wvRef} style={{height: '100%'}} src={this.props.src}/> */}
        </div>
      </Drawer>
    )
  }
}

export default WebviewDrawer
