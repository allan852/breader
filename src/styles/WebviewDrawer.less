@import './Values.less';

.webview-drawer {
  .ant-drawer-body {
    position: relative;
    background-color: @grey-background-color;
    padding: 42px 0 0 0;
    height: 100%;
  }
  .drawer-header {
    .wrapper-header;
    .anticon {
      color: @black-text-color;
      &:hover {
        color: @black-text-color-light;
      }
    }
  }
  .drawer-header-left {
    .wrapper-hf-left;
    .anticon {
      margin-top: 13px;
      margin-left: 16px;
    }
  }
  .drawer-header-right {
    .wrapper-hf-right;
    .anticon {
      margin-top: 13px;
      margin-right: 16px;
    }
  }
  .drawer-content {
    position: relative;
    background-size: contain;
    // background-image: url(../images/book.gif);
    background-repeat: no-repeat;
    background-position: center;
    width: 100%;
    height: 100%;
  }
  .webview-progress {
    position: absolute;
    top: 40px;
    right: 0;
    left: 0;
    line-height: 3px;
    z-index: 99;
    .ant-progress-bg {
      border-radius: 0px !important;
    }
    .ant-progress-inner {
      background-color: @none-color;
    }
  }
  webview {
    height: 100%;
  }
}
