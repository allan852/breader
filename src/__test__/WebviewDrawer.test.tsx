import Enzyme from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import EnzymeToJson from 'enzyme-to-json'
import React from 'react'
import WebviewDrawer, { IWebviewDrawerProps } from '../components/WebviewDrawer'

Enzyme.configure({ adapter: new EnzymeAdapter() })

describe('WebviewDrawer Testing', () => {
    const propsMock: IWebviewDrawerProps = {
        onClose: jest.fn(),
        src: 'https://github.com',
        visible: true,
        width: 100,
    }

    it('dom testing', () => {
        const component = Enzyme.shallow(
            <WebviewDrawer {...propsMock} />
        )
    })

    it('snapshot testing', () => {
        const component = Enzyme.mount(<WebviewDrawer {...propsMock} />)
        const tree = EnzymeToJson(component)
        expect(tree).toMatchSnapshot()
    })
})
