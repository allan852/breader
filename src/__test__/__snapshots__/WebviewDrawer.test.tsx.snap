// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebviewDrawer Testing snapshot testing 1`] = `
<WebviewDrawer
  onClose={[MockFunction]}
  src="https://github.com"
  visible={true}
  width={100}
>
  <withConfigConsumer(Drawer)
    className="webview-drawer"
    closable={false}
    onClose={[Function]}
    visible={true}
    width={100}
  >
    <Drawer
      className="webview-drawer"
      closable={false}
      getPrefixCls={[Function]}
      height={256}
      level={null}
      mask={true}
      maskClosable={true}
      onClose={[Function]}
      placement="right"
      prefixCls="ant-drawer"
      renderEmpty={[Function]}
      visible={true}
      width={100}
    >
      <Drawer
        afterVisibleChange={[Function]}
        className="webview-drawer"
        closable={false}
        duration=".3s"
        ease="cubic-bezier(0.78, 0.14, 0.15, 0.86)"
        firstEnter={false}
        getContainer="body"
        getPrefixCls={[Function]}
        handler={false}
        keyboard={true}
        level={null}
        mask={true}
        maskClosable={true}
        maskStyle={Object {}}
        onChange={[Function]}
        onClose={[Function]}
        onMaskClick={[Function]}
        open={true}
        placement="right"
        prefixCls="ant-drawer"
        renderEmpty={[Function]}
        showMask={true}
        style={
          Object {
            "transform": undefined,
            "zIndex": undefined,
          }
        }
        visible={true}
        width={100}
        wrapperClassName=""
      >
        <Portal
          containerInfo={
            <div>
              <div
                class="ant-drawer ant-drawer-right ant-drawer-open webview-drawer"
                tabindex="-1"
              >
                <div
                  class="ant-drawer-mask"
                />
                <div
                  class="ant-drawer-content-wrapper"
                  style="width: 100px;"
                >
                  <div
                    class="ant-drawer-content"
                  >
                    <div
                      class="ant-drawer-wrapper-body"
                      style="overflow: auto; height: 100%;"
                    >
                      <div
                        class="ant-drawer-body"
                      >
                        <div
                          class="drawer-header"
                        >
                          <div
                            class="drawer-header-left"
                          >
                            <i
                              aria-label="icon: close"
                              class="anticon anticon-close"
                            >
                              <svg
                                aria-hidden="true"
                                class=""
                                data-icon="close"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                />
                              </svg>
                            </i>
                          </div>
                          <div
                            class="drawer-header-right"
                          >
                            <i
                              aria-label="icon: compass"
                              class="anticon anticon-compass"
                            >
                              <svg
                                aria-hidden="true"
                                class=""
                                data-icon="compass"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm198.4-588.1a32 32 0 0 0-24.5.5L414.9 415 296.4 686c-3.6 8.2-3.6 17.5 0 25.7 3.4 7.8 9.7 13.9 17.7 17 3.8 1.5 7.7 2.2 11.7 2.2 4.4 0 8.7-.9 12.8-2.7l271-118.6 118.5-271a32.06 32.06 0 0 0-17.7-42.7zM576.8 534.4l26.2 26.2-42.4 42.4-26.2-26.2L380 644.4 447.5 490 422 464.4l42.4-42.4 25.5 25.5L644.4 380l-67.6 154.4zM464.4 422L422 464.4l25.5 25.6 86.9 86.8 26.2 26.2 42.4-42.4-26.2-26.2-86.8-86.9z"
                                />
                              </svg>
                            </i>
                          </div>
                        </div>
                        <div
                          class="drawer-content"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          }
        >
          <div
            className="ant-drawer ant-drawer-right ant-drawer-open webview-drawer"
            onKeyDown={[Function]}
            onTransitionEnd={[Function]}
            style={
              Object {
                "transform": undefined,
                "zIndex": undefined,
              }
            }
            tabIndex={-1}
          >
            <div
              className="ant-drawer-mask"
              onClick={[Function]}
              style={Object {}}
            />
            <div
              className="ant-drawer-content-wrapper"
              style={
                Object {
                  "height": undefined,
                  "msTransform": "",
                  "transform": "",
                  "width": "100px",
                }
              }
            >
              <div
                className="ant-drawer-content"
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                <div
                  className="ant-drawer-wrapper-body"
                  onTransitionEnd={[Function]}
                  style={
                    Object {
                      "height": "100%",
                      "overflow": "auto",
                    }
                  }
                >
                  <div
                    className="ant-drawer-body"
                  >
                    <div
                      className="drawer-header"
                    >
                      <div
                        className="drawer-header-left"
                        onClick={[Function]}
                      >
                        <Icon
                          type="close"
                        >
                          <LocaleReceiver
                            componentName="Icon"
                          >
                            <i
                              aria-label="icon: close"
                              className="anticon anticon-close"
                            >
                              <IconReact
                                className=""
                                type="close-o"
                              >
                                <svg
                                  aria-hidden="true"
                                  className=""
                                  data-icon="close"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  key="svg-close"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                    key="svg-close-svg-0"
                                  />
                                </svg>
                              </IconReact>
                            </i>
                          </LocaleReceiver>
                        </Icon>
                      </div>
                      <div
                        className="drawer-header-right"
                        onClick={[Function]}
                      >
                        <Icon
                          type="compass"
                        >
                          <LocaleReceiver
                            componentName="Icon"
                          >
                            <i
                              aria-label="icon: compass"
                              className="anticon anticon-compass"
                            >
                              <IconReact
                                className=""
                                type="compass-o"
                              >
                                <svg
                                  aria-hidden="true"
                                  className=""
                                  data-icon="compass"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  key="svg-compass"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm198.4-588.1a32 32 0 0 0-24.5.5L414.9 415 296.4 686c-3.6 8.2-3.6 17.5 0 25.7 3.4 7.8 9.7 13.9 17.7 17 3.8 1.5 7.7 2.2 11.7 2.2 4.4 0 8.7-.9 12.8-2.7l271-118.6 118.5-271a32.06 32.06 0 0 0-17.7-42.7zM576.8 534.4l26.2 26.2-42.4 42.4-26.2-26.2L380 644.4 447.5 490 422 464.4l42.4-42.4 25.5 25.5L644.4 380l-67.6 154.4zM464.4 422L422 464.4l25.5 25.6 86.9 86.8 26.2 26.2 42.4-42.4-26.2-26.2-86.8-86.9z"
                                    key="svg-compass-svg-0"
                                  />
                                </svg>
                              </IconReact>
                            </i>
                          </LocaleReceiver>
                        </Icon>
                      </div>
                    </div>
                    <div
                      className="drawer-content"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </Portal>
      </Drawer>
    </Drawer>
  </withConfigConsumer(Drawer)>
</WebviewDrawer>
`;
